eessi/testsuite/tests/apps/QuantumESPRESSO.py

```diff
@@ -0,0 +1,93 @@
+
+"concrete test cases": all test combinations that will actually run:
+- test variants
+- valid system:partition+programming environment combinations
+https://reframe-hpc.readthedocs.io/en/stable/tutorial_deps.html#listing-dependencies
+
+Tests can be filtered by name, tag, programming environment, system, partition, or maintainer,
+see https://reframe-hpc.readthedocs.io/en/stable/manpage.html#test-filtering
+
+Hooks acting on all possible test combinations (before filtering) are called after the 'init' stage.
+Hooks acting on concrete test cases (after filtering) are called after the 'setup' stage.
+
+See also https://reframe-hpc.readthedocs.io/en/stable/pipeline.html
+"""
+
+import reframe as rfm
+from hpctestlib.sciapps.qespresso.benchmarks import QEspressoPWCheck
+from reframe.core.builtins import (  # added only to make the linter happy
+    parameter, run_after)
+
+from eessi.testsuite import hooks
+from eessi.testsuite.constants import SCALES, TAGS
+from eessi.testsuite.utils import find_modules, log
+
+
+@rfm.simple_test
+class EESSI_QuantumESPRESSO_PW(QEspressoPWCheck):
+    scale = parameter(SCALES.keys())
+    valid_prog_environs = ['default']
+    valid_systems = ['*']
+    time_limit = '30m'
+    module_name = parameter(find_modules('QuantumESPRESSO'))
+
+    @run_after('init')
+    def run_after_init(self):
+        """Hooks to run after the init phase"""
+
+        # Filter on which scales are supported by the partitions defined in the ReFrame configuration
+        hooks.filter_supported_scales(self)
+
+        # Make sure that GPU tests run in partitions that support running on a GPU,
+        # and that CPU-only tests run in partitions that support running CPU-only.
+        # Also support setting valid_systems on the cmd line.
+        hooks.filter_valid_systems_by_device_type(self, required_device_type=self.nb_impl)
+
+        # Support selecting modules on the cmd line.
+        hooks.set_modules(self)
+
+        # Support selecting scales on the cmd line via tags.
+        hooks.set_tag_scale(self)
+
+    @run_after('init')
+    def set_tag_ci(self):
+        """Set tag CI on smallest benchmark, so it can be selected on the cmd line via --tag CI"""
+        min_ecut = min(QEspressoPWCheck.ecut.values)
+        min_nbnd = min(QEspressoPWCheck.nbnd.values)
+        if self.ecut == min_ecut and self.nbnd == min_nbnd:
+            self.tags.add(TAGS['CI'])
+            log(f'tags set to {self.tags}')
+
+    @run_after('setup')
+    def run_after_setup(self):
+        """Hooks to run after the setup phase"""
+
+        # Calculate default requested resources based on the scale:
+        # 1 task per CPU for CPU-only tests, 1 task per GPU for GPU tests.
+        # Also support setting the resources on the cmd line.
+        hooks.assign_tasks_per_compute_unit(test=self, compute_unit=self.nb_impl)
+
+    @run_after('setup')
+    def set_omp_num_threads(self):
```

Comment:
Note that setting `OMP_NUM_THREADS` was very important for GROMACS, since it fails if an `OMP_NUM_THREADS` was already set that doesn't match the thread count requested on command line. Of course, I don't know how QE behaves and if it is important to set `OMP_NUM_THREADS`. Note that with:
```
hooks.assign_tasks_per_compute_unit(test=self, compute_unit=COMPUTE_UNIT[CPU]
```
You'd get one process per CPU core anyway, so that would basically limit QE to one thread per process.

Reply:
It is, to my experience if `OMP_NUM_THREADS` is not set and running QE with (where XXX is the max number of procs)

    mpirun -n XXX pw.x ....

it is actually possible that QE will start with XXX procs and XXX threads resulting in a very likely crashed or super slow run

Reply:
The batch script looks fine, one thing that might happen since we are running with as much as 192 cores is that the smallest benchmark might fail due to QE not being able to assign cores, in that case we might want to run the biggest test only (`ecut = 150`  `nbnd = 200` on the large partitions)

https://github.com/reframe-hpc/reframe/blob/fe1c3b406ffbd4c020e47486ed98f238a734cc8f/hpctestlib/sciapps/qespresso/benchmarks.py#L67

Reply:
> it is actually possible that QE will start with XXX procs and XXX threads resulting in a very likely crashed or super slow run

You mean that on a e.g. 128-core node, it would start 128 processes, with 128 threads _each_? Ok, then lets keep it. As said, it doesn't hurt :)

> The batch script looks fine, one thing that might happen since we are running with as much as 192 cores is that the smallest benchmark might fail due to QE not being able to assign cores, in that case we might want to run the biggest test only (ecut = 150 nbnd = 200 on the large partitions)

Ok. We do have a bit of a catch-22 with ReFrame in these situations: the core count (for a single node) is only known in the `setup` phase, because only then is the `current_partition` set and can we query the amount of cores for the current partition (see e.g. [here](https://github.com/EESSI/test-suite/blob/ede05e414922c577eda954e6ef6c67abd2b9f233/eessi/testsuite/hooks.py#L88)). However, we can only prevent test instances from being generated _before_ the `setup` phase. 

So, we cannot _not_ generate a test instance. The only thing we can really do here, is skip the test instance if a certain condition becomes true. You can find an example [here](https://github.com/EESSI/test-suite/blob/ede05e414922c577eda954e6ef6c67abd2b9f233/eessi/testsuite/tests/apps/osu.py#L148) which is quite similar: that OSU test requires at least 2 GPU devices to run, but if the current partition only happens to have 1 GPU per node, we know the test isn't valid for such a configuration.

In your case it would be something like
```
self.skip_if(self.ecut < XYZ and self.num_tasks > ABC, f"Skipping test with ecut={self.ecut} and num_tasks={self.num_tasks} because the problem size is too small for the number of tasks")
```

Reply:
I am gonna check if i can find a way to simulate running with 192 cores and see if the test will run or not and adjust accordingly